rust/kernel/spi.rs

```diff
@@ -0,0 +1,279 @@
+
+//! This module provides safer and higher level abstraction over the kernel's SPI types
+//! and functions.
+//!
+//! C header: [`include/linux/spi/spi.h`](../../../../include/linux/spi/spi.h)
+
+use crate::bindings;
+use crate::c_types;
+use crate::error::{Error, Result};
+use crate::str::CStr;
+use alloc::boxed::Box;
+use core::pin::Pin;
+
+/// Wrapper struct around the kernel's `spi_device`
+#[derive(Clone, Copy)]
+pub struct SpiDevice(*mut bindings::spi_device);
+
+impl SpiDevice {
+    /// Create an SpiDevice from a mutable spi_device raw pointer. This function is unsafe
+    /// as the pointer might be invalid.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be valid. This can be achieved by calling `to_ptr` on a previously
+    /// constructed, safe SpiDevice instance, or by making sure that the pointer points
+    /// to valid memory.
+    /// You probably do not want to use this abstraction directly. It is mainly used
+    /// by this abstraction to wrap valid pointers given by the Kernel to the different
+    /// spi methods: probe, remove and shutdown
+    pub unsafe fn from_ptr(dev: *mut bindings::spi_device) -> Self {
+        SpiDevice(dev)
+    }
+
+    /// Access the raw pointer from an SpiDevice instance
+    pub fn to_ptr(&mut self) -> *mut bindings::spi_device {
+        self.0
+    }
+}
+
+/// Registration of an SPI driver
+pub struct DriverRegistration {
+    this_module: &'static crate::ThisModule,
+    registered: bool,
+    name: &'static CStr,
+    spi_driver: bindings::spi_driver,
+}
+
+/// Represents which methods of the SPI driver shall be implemented
+pub struct ToUse {
+    /// The `probe` field of `spi_driver`
+    pub probe: bool,
+
+    /// The `remove` field of `spi_driver`
+    pub remove: bool,
+
+    /// The `shutdown` field of `spi_driver`
+    pub shutdown: bool,
+}
+
+/// Default table to use for SPI methods. All fields are set to false, meaning that the
+/// kernel's default implementations will be used.
+pub const USE_NONE: ToUse = ToUse {
+    probe: false,
+    remove: false,
+    shutdown: false,
+};
+
+/// Corresponds to the kernel's spi_driver's methods. Implement this trait on a type to
+/// express the need of a custom probe, remove or shutdown function for your SPI driver.
+/// Use the [`declare_spi_methods`] macro to declare which methods you wish to use.
+pub trait SpiMethods {
+    /// The methods to define. Use [`declare_spi_methods`] to declare them
+    const TO_USE: ToUse;
+
+    /// Corresponds to the kernel's spi_driver's probe method field
+    fn probe(mut _spi_dev: SpiDevice) -> Result {
+        Ok(())
+    }
+
+    /// Corresponds to the kernel's spi_driver's remove method field
+    fn remove(mut _spi_dev: SpiDevice) -> Result {
+        Ok(())
+    }
+
+    /// Corresponds to the kernel's spi_driver's shutdown method field
+    fn shutdown(mut _spi_dev: SpiDevice) {}
+}
+
+/// Populate the TO_USE field in the [`SpiMethods`] implementer
+///
+/// ```rust
```

Comment:
I think `rust` is the default (I might be wrong though, please check if the output is highlighted).

```suggestion
/// ```
```

Reply:
Yes, rust is the default. That was overzealous on our part, sorry!

Reply:
Fixed by 92a19c2ef2d0d0780c34327b273a21db627af711